robot/impl/local_robot.go

```diff
@@ -1067,6 +1050,19 @@ func (r *localRobot) Reconfigure(ctx context.Context, newConfig *config.Config)
 	// Set mostRecentConfig if resources were not equal.
 	r.mostRecentCfg = *newConfig
 
+	// We need to pre-add the new modules so that resource validation can check against the new models
+	// TODO(RSDK-4383) These lines are taken from uppdateResources() and should be refactored as part of this bugfix
+	for _, mod := range diff.Added.Modules {
+		if err := mod.Validate(""); err != nil {
+			r.manager.logger.Errorw("module config validation error; skipping", "module", mod.Name, "error", err)
+			continue
+		}
+		if err := r.manager.moduleManager.Add(ctx, mod); err != nil {
+			r.manager.logger.Errorw("error adding module", "module", mod.Name, "error", err)
+			continue
+		}
+	}
```

Comment:
Talked with @Otterverse about these lines on Friday.

These lines highlight a longstanding bug in rdk that currently exists in rdk@head: [Resource configs are not re-validated after a changed executable path of the underlying module](https://viam.atlassian.net/browse/RSDK-4383)

James and I are divided as to whether this bug should block this PR or should be a `priority:High` to fix soon after this is merged.


(These lines are also ugly and probably shouldn't be here, but they preserve almost exactly the same behavior as local_robot.go used to have. Fixing this ugliness should be done very soon (possibly as part of RSDK-4383))

Reply:
Yeah I remember these lines being key in `newWithResources` to allow resources to both be validated by and added to the correct modules initially; doesn't surprise me they're still needed here. I do think [Resource configs are not re-validated after a changed executable path of the underlying module](https://viam.atlassian.net/browse/RSDK-4383) is a bug and an important one, but my sense is that it's out of scope for this PR. I'm also not sure RSDK-4383 will be able to "get rid" of these lines; I'm suspicious of a solution where modular resources cannot validate nor build until their module _happens_ to be added. I think we should ensure that modules are started up before their resources.